File: pyrevitlib/pyrevit/coreutils/envvars.py
"""pyRevit managed environment variables framework.

pyRevit provides the environment variables framework to the pyRevit core
and all pyRevit tools so they can store arbitary data withing the running host
session and share small data quickly between script runs.

Some settings needs to be set for the current session and might need to affect
the behaviour of all individual scripts inside the extensions.
(e.g. If user activates the ``DEBUG`` mode, all scripts should follow and log
the debug entries.) The information is saved using ``AppDomain.GetData`` and
``SetData`` in a dictionary parameter. The dictionary is used to minimize the
addition of named parameters to the AppDomain. The dictionary then includes
all the internal parameters and their associated value. This way each script
does not need to read the usersettings data which reduces io and saves time.


pyRevit uses environment variables extensively at its core and making changes
to the core environment variables (starting with ``PYREVIT_``) through
scripts is strongly prohibited.

Examples:
    ```python
    from pyrevit.coreutils import envvars
    envvars.set_pyrevit_env_var('MY_SCRIPT_STATUS', True)
    envvars.set_pyrevit_env_var('MY_SCRIPT_CONFIG', {'someconfig': True})
    ```

    Then another script or same script when executed later within the same
    session can query the shared environment variable:

    ```python
    envvars.get_pyrevit_env_vars('MY_SCRIPT_STATUS')
    ```
    True
    ```python
    envvars.get_pyrevit_env_vars('MY_SCRIPT_CONFIG')
    ```
    {'someconfig': True}
"""

from pyrevit.framework import AppDomain
from pyrevit.labs import Common

PRODUCT_NAME = str(Common.PyRevitLabsConsts.ProductName).upper()
# root env var dictionary key.
# must be the same in this file and pyrevit/loader/runtime/envvars.cs
ENV_VAR_DICT_NAME = PRODUCT_NAME + "EnvVarsDict"

SESSIONUUID_ENVVAR = PRODUCT_NAME + "_UUID"
APPVERSION_ENVVAR = PRODUCT_NAME + "_APPVERSION"
VERSION_ENVVAR = PRODUCT_NAME + "_VERSION"
CLONENAME_ENVVAR = PRODUCT_NAME + "_CLONE"
IPYVERSION_ENVVAR = PRODUCT_NAME + "_IPYVERSION"
CPYVERSION_ENVVAR = PRODUCT_NAME + "_CPYVERSION"

LOGGING_LEVEL_ENVVAR = PRODUCT_NAME + "_LOGGINGLEVEL"
FILELOGGING_ENVVAR = PRODUCT_NAME + "_FILELOGGING"

LOADEDASSMS_ENVVAR = PRODUCT_NAME + "_LOADEDASSMS"
REFEDASSMS_ENVVAR = PRODUCT_NAME + "_REFEDASSMS"

TELEMETRYSTATE_ENVVAR = PRODUCT_NAME + "_TELEMETRYSTATE"
TELEMETRYUTCTIMESTAMPS_ENVVAR = PRODUCT_NAME + "_TELEMETRYUTCTIMESTAMPS"
TELEMETRYDIR_ENVVAR = PRODUCT_NAME + "_TELEMETRYDIR"
TELEMETRYFILE_ENVVAR = PRODUCT_NAME + "_TELEMETRYFILE"
TELEMETRYSERVER_ENVVAR = PRODUCT_NAME + "_TELEMETRYSERVER"
TELEMETRYINCLUDEHOOKS_ENVVAR = PRODUCT_NAME + "_TELEMETRYINCLUDEHOOKS"

APPTELEMETRYSTATE_ENVVAR = PRODUCT_NAME + "_APPTELEMETRYSTATE"
APPTELEMETRYHANDLER_ENVVAR = \
    PRODUCT_NAME + "_APPTELEMETRYHANDLER"
APPTELEMETRYSERVER_ENVVAR = \
    PRODUCT_NAME + "_APPTELEMETRYSERVER"
APPTELEMETRYEVENTFLAGS_ENVVAR = \
    PRODUCT_NAME + "_APPTELEMETRYEVENTFLAGS"

ROUTES_SERVER = PRODUCT_NAME + "_ROUTESSERVER"
ROUTES_ROUTES = PRODUCT_NAME + "_ROUTESROUTES"

HOOKS_ENVVAR = PRODUCT_NAME + "_HOOKS"
HOOKSHANDLER_ENVVAR = PRODUCT_NAME + "_HOOKSHANDLER"

AUTOUPDATING_ENVVAR = PRODUCT_NAME + "_AUTOUPDATE"
OUTPUT_STYLESHEET_ENVVAR = PRODUCT_NAME + "_STYLESHEET"
RIBBONUPDATOR_ENVVAR = PRODUCT_NAME + "_RIBBONUPDATOR"
TABCOLORIZER_ENVVAR = PRODUCT_NAME + "_TABCOLORIZER"


def get_pyrevit_env_vars():
    """Get the root dictionary, holding all environment variables."""
    return AppDomain.CurrentDomain.GetData(ENV_VAR_DICT_NAME)


def get_pyrevit_env_var(param_name):
    """Get value of a parameter shared between all scripts.

    Args:
        param_name (str): name of environment variable

    Returns:
        (object): any object stored as the environment variable value
    """
    # This function returns None if it can not find the parameter.
    # Thus value of None should not be used for params

    data_dict = get_pyrevit_env_vars()
    return data_dict.get(param_name) if data_dict else None


def set_pyrevit_env_var(param_name, param_value):
    """Set value of a parameter shared between all scripts.

    Args:
        param_name (str): name of environment variable
        param_value (object): any python object
    """
    # Get function returns None if it can not find the parameter.
    # Thus value of None should not be used for params
    data_dict = get_pyrevit_env_vars() or {}
    data_dict[param_name] = param_value
    AppDomain.CurrentDomain.SetData(ENV_VAR_DICT_NAME, data_dict)

File: pyrevitlib/pyrevit/loader/basetypes/envvars.cs
using System;
using System.IO;
using System.Collections.Generic;
using IronPython.Runtime;


namespace PyRevitBaseClasses
{
    public static class EnvDictionaryKeys
    {
        public static string keyPrefix = "PYREVIT";

        public static string sessionUUID = string.Format("{0}_UUID", keyPrefix);
        public static string RevitVersion = string.Format("{0}_APPVERSION", keyPrefix);
        public static string pyRevitVersion = string.Format("{0}_VERSION", keyPrefix);
        public static string pyRevitClone = string.Format("{0}_CLONE", keyPrefix);
        public static string pyRevitIpyVersion = string.Format("{0}_IPYVERSION", keyPrefix);
        public static string pyRevitCpyVersion = string.Format("{0}_CPYVERSION", keyPrefix);

        public static string loggingLevel = string.Format("{0}_LOGGINGLEVEL", keyPrefix);
        public static string fileLogging = string.Format("{0}_FILELOGGING", keyPrefix);

        public static string loadedAssm = string.Format("{0}_LOADEDASSMS", keyPrefix);
        public static string refedAssms = string.Format("{0}_REFEDASSMS", keyPrefix);

        public static string telemetryState = string.Format("{0}_TELEMETRYSTATE", keyPrefix);
        public static string telemetryFileDir = string.Format("{0}_TELEMETRYDIR", keyPrefix);
        public static string telemetryFilePath = string.Format("{0}_TELEMETRYFILE", keyPrefix);
        public static string telemetryServerUrl = string.Format("{0}_TELEMETRYSERVER", keyPrefix);
        public static string appTelemetryState = string.Format("{0}_APPTELEMETRYSTATE", keyPrefix);
        public static string appTelemetryHandler = string.Format("{0}_APPTELEMETRYHANDLER", keyPrefix);
        public static string appTelemetryServerUrl = string.Format("{0}_APPTELEMETRYSERVER", keyPrefix);
        public static string appTelemetryEventFlags = string.Format("{0}_APPTELEMETRYEVENTFLAGS", keyPrefix);

        public static string hooks = string.Format("{0}_HOOKS", keyPrefix);
        public static string hooksHandler = string.Format("{0}_HOOKSHANDLER", keyPrefix);

        public static string autoupdating = string.Format("{0}_AUTOUPDATING", keyPrefix);
        public static string outputStyleSheet = string.Format("{0}_STYLESHEET", keyPrefix);

    }

    public class EnvDictionary
    {
        private PythonDictionary _envData = null;

        public string SessionUUID;
        public string RevitVersion;
        public string PyRevitVersion;
        public string PyRevitClone;
        public int PyRevitIPYVersion;
        public int PyRevitCPYVersion;

        public int LoggingLevel;
        public bool FileLogging;

        public string[] LoadedAssemblies;
        public string[] ReferencedAssemblies;

        public bool TelemetryState;
        public string TelemetryFilePath;
        public string TelemetryServerUrl;

        public bool AppTelemetryState;
        public string AppTelemetryServerUrl;
        public string AppTelemetryEventFlags;

        public Dictionary<string, Dictionary<string, string>> EventHooks = 
            new Dictionary<string, Dictionary<string, string>>();

        public string ActiveStyleSheet;
        public bool AutoUpdate;


        public EnvDictionary()
        {
            // get the dictionary from appdomain
            _envData = (PythonDictionary)AppDomain.CurrentDomain.GetData(DomainStorageKeys.pyRevitEnvVarsDictKey);

            // base info
            if (_envData.Contains(EnvDictionaryKeys.sessionUUID))
                SessionUUID = (string)_envData[EnvDictionaryKeys.sessionUUID];

            if (_envData.Contains(EnvDictionaryKeys.RevitVersion))
                RevitVersion = (string)_envData[EnvDictionaryKeys.RevitVersion];

            if (_envData.Contains(EnvDictionaryKeys.pyRevitVersion))
                PyRevitVersion = (string)_envData[EnvDictionaryKeys.pyRevitVersion];

            if (_envData.Contains(EnvDictionaryKeys.pyRevitClone))
                PyRevitClone = (string)_envData[EnvDictionaryKeys.pyRevitClone];

            if (_envData.Contains(EnvDictionaryKeys.pyRevitIpyVersion))
                PyRevitIPYVersion = (int)_envData[EnvDictionaryKeys.pyRevitIpyVersion];

            if (_envData.Contains(EnvDictionaryKeys.pyRevitCpyVersion))
                PyRevitCPYVersion = (int)_envData[EnvDictionaryKeys.pyRevitCpyVersion];

            // logging
            if (_envData.Contains(EnvDictionaryKeys.loggingLevel))
                LoggingLevel = (int)_envData[EnvDictionaryKeys.loggingLevel];
            if (_envData.Contains(EnvDictionaryKeys.fileLogging))
                FileLogging = (bool)_envData[EnvDictionaryKeys.fileLogging];

            // assemblies
            if (_envData.Contains(EnvDictionaryKeys.loadedAssm))
                LoadedAssemblies = ((string)_envData[EnvDictionaryKeys.loadedAssm]).Split(Path.PathSeparator);
            if (_envData.Contains(EnvDictionaryKeys.refedAssms))
                ReferencedAssemblies = ((string)_envData[EnvDictionaryKeys.refedAssms]).Split(Path.PathSeparator);

            // script telemetry
            if (_envData.Contains(EnvDictionaryKeys.telemetryState))
                TelemetryState = (bool)_envData[EnvDictionaryKeys.telemetryState];

            if (_envData.Contains(EnvDictionaryKeys.telemetryFilePath))
                TelemetryFilePath = (string)_envData[EnvDictionaryKeys.telemetryFilePath];

            if (_envData.Contains(EnvDictionaryKeys.telemetryServerUrl))
                TelemetryServerUrl = (string)_envData[EnvDictionaryKeys.telemetryServerUrl];

            // app events telemetry
            if (_envData.Contains(EnvDictionaryKeys.appTelemetryState))
                AppTelemetryState = (bool)_envData[EnvDictionaryKeys.appTelemetryState];

            if (_envData.Contains(EnvDictionaryKeys.appTelemetryServerUrl))
                AppTelemetryServerUrl = (string)_envData[EnvDictionaryKeys.appTelemetryServerUrl];

            if (_envData.Contains(EnvDictionaryKeys.appTelemetryEventFlags))
                AppTelemetryEventFlags = (string)_envData[EnvDictionaryKeys.appTelemetryEventFlags];

            // hooks
            if (_envData.Contains(EnvDictionaryKeys.hooks))
                EventHooks = (Dictionary<string, Dictionary<string, string>>)_envData[EnvDictionaryKeys.hooks];
            else
                _envData[EnvDictionaryKeys.hooks] = EventHooks;

            // misc
            if (_envData.Contains(EnvDictionaryKeys.autoupdating))
                AutoUpdate = (bool)_envData[EnvDictionaryKeys.autoupdating];

            if (_envData.Contains(EnvDictionaryKeys.outputStyleSheet))
                ActiveStyleSheet = (string)_envData[EnvDictionaryKeys.outputStyleSheet];
        }

        public void ResetEventHooks() {
            ((Dictionary<string, Dictionary<string, string>>)_envData[EnvDictionaryKeys.hooks]).Clear();
        }
    }
}
